### PR TITLE
Change API to return LogicalPlan instead of DataFrame

### PR DESCRIPTION
**Pull request by andygrove**
As suggested by @nseekhao, given that `to_substrait_rel()` takes `LogicalPlan` and returns Substrait `Rel`, it would be better if `from_subtrait_rel()` takes Substrait `Rel` and returns `LogicalPlan` rather than `DataFrame`.

### CODE REVIEW

File: src/consumer.rs [hunk 1]
```diff
@@ -309,9 +316,10 @@ pub async fn from_substrait_rel(
                     }
                 })
                 .collect::<Result<Vec<_>>>()?;
-            let left_cols: Vec<&str> = pairs.iter().map(|(l, _)| l.as_str()).collect();
-            let right_cols: Vec<&str> = pairs.iter().map(|(_, r)| r.as_str()).collect();
-            left.join(right, join_type, &left_cols, &right_cols, None)
+            let left_cols: Vec<Column> = pairs.iter().map(|(l, _)| l.clone()).collect();
+            let right_cols: Vec<Column> = pairs.iter().map(|(_, r)| r.clone()).collect();
```
Comment (nseekhao) [inline src/consumer.rs:320]: Would this also work?

```Rust
let (left_cols, right_cols): (Vec<_>, Vec<_>) = pairs.iter().cloned().unzip();
```

File: src/consumer.rs [hunk 2]
```diff
@@ -309,9 +316,10 @@ pub async fn from_substrait_rel(
...
+            left.join(&right, join_type, (left_cols, right_cols), None)?
```
Comment (nseekhao) [inline src/consumer.rs:321]: Is the last parameter (`filter`) that's set to `None` the same field semantically as Substrait `JoinRel`'s  [post_join_filter](https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L161)?

If not, please ignore the comment.

If so, should we check if the `JoinRel` has `post_join_filter`? And if so, parse it?

File: tests/roundtrip.rs [hunk 3]
```diff
@@ -174,9 +173,8 @@ mod tests {
         let df = ctx.sql(sql).await?;
         let plan1 = df.to_logical_plan()?;
         let proto = to_substrait_plan(&plan1)?;
-
-        let df = from_substrait_plan(&mut ctx, &proto).await?;
-        let plan2 = df.to_logical_plan()?;
+        let plan2 = from_substrait_plan(&mut ctx, &proto).await?;
+        let plan2 = ctx.optimize(&plan2)?;
```
Comment (nseekhao) [inline tests/roundtrip.rs:177]: Clarification question. Why do we need to call `optimize()` here?

Comment (andygrove) [reply tests/roundtrip.rs:177]: Before the changes in this PR, we were receiving a `DataFrame` and then calling `DataFrame::to_logical_plan`, which performed the optimization step, so the existing tests were expecting an optimized plan.

File: src/consumer.rs (hunk 1 above)
Comment (andygrove) [reply src/consumer.rs]: Thanks. Yes, that is much nicer. Applied.

File: tests/roundtrip.rs (hunk 3 above)
Comment (nseekhao) [reply tests/roundtrip.rs:177]: Ah I see. Thank you for the explanation!

File: src/consumer.rs (hunk 2 above)
Comment (andygrove) [reply src/consumer.rs:320]: Yes, I think those probably are the same concept. It would be good to implement this in a separate PR.

**Review by waynexia — APPROVED** [review]
Looks good to me. By the way, I prefer to remove `optimize()` calls later as it is out of the scope of serializing/deserializing. The downstream consumer can do optimization by themselves.

**Comment by andygrove** [comment]
> Looks good to me. By the way, I prefer to remove `optimize()` calls later as it is out of the scope of serializing/deserializing. The downstream consumer can do optimization by themselves.

Yes, I agree with that.

**Pull request merged**